### PR TITLE
[WIP] Separate basic operations into data type based structures

### PR DESCRIPTION
**Pull request by xinrong-databricks**
## Refactored binary operators
`__add__`, `__sub__`, `__mul__`, `__truediv__`, `__floordiv__`, `__pow__`, `__mod__`,
`__radd__`, `__rsub__`, `__rmul__`, `__rtruediv__`, `__rfloordiv__`, `__rpow__`,`__rmod__`

## Existing behaviors are preserved; places where new logic introduced
Error messages about `__pow__`, `__mod__`,
Error messages of basic operations against CategoricalOps
DateOps
Error message for the non-numeric
Arithmetic operations for bool series

## Tests
Tests are introduced under databricks.koalas.tests.data_type_ops. One test file per DataTypeOps class.

## Note

- Inconsistent TypeError messages are expected since we want to preserve the existing behavior to ensure a smooth refactoring.TypeError messages should be consolidated in a follow-up PR.

- We have some tests of arithmetic operations in

      databricks/koalas/tests/test_ops_on_diff_frames.py
      databricks/koalas/tests/test_dataframe.py
      databricks/koalas/tests/test_series.py

They are preserved to ensure refactored code respects the existing behavior. And they should be consolidated in a follow-up PR.

- BooleanOps (arithmetic)

- ExtentionDtypeOps

### CODE REVIEW

File: databricks/koalas/data_type_ops.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
+import datetime
```
Comment (ueshin) [inline databricks/koalas/data_type_ops.py:1]: License header?

File: databricks/koalas/data_type_ops.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import decimal
+import numpy as np
+
+from pandas.api.types import CategoricalDtype
+
+
+class DataTypeOps:
+    def __new__(cls, dtype, spark_type):
```
Comment (ueshin) [inline databricks/koalas/data_type_ops.py:9]: Shall we only allow `dtype: Dtype`?

File: databricks/koalas/data_type_ops.py (hunk 1 above)
Comment (xinrong-databricks) [reply databricks/koalas/data_type_ops.py:20]: Yeah added.

File: databricks/koalas/data_type_ops.py (hunk 2 above)
Comment (xinrong-databricks) [reply databricks/koalas/data_type_ops.py]: Added type annotations. I'm wondering what is `spark_type`'s usage in `__new__`?

Comment (ueshin) [reply databricks/koalas/data_type_ops.py]: We can't decide the actual type class without it.
For example, without it, how do you decide the actual class if dtype is `np.dtype('object')`?

Comment (xinrong-databricks) [reply databricks/koalas/data_type_ops.py]: Got it, thanks!

File: databricks/koalas/data_type_ops/num_ops.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+class FractionalOps(NumericOps):
+    """
+    The class for binary operations of Koalas objects with spark types: FloatType and DoubleType.
+    """
+
+    pass
+
+
+class DecimalOps(FractionalOps):
```
Comment (xinrong-databricks) [inline databricks/koalas/data_type_ops/num_ops.py:192]: @ueshin Would you show examples of why we wanted to make `FractionalOps` and `DecimalOps` each an individual class? Thanks!

Comment (ueshin) [reply databricks/koalas/data_type_ops/num_ops.py:192]: `DoubleType` and `FloatType` could have `NaN`, whereas `DecimalType` not.

Comment (xinrong-databricks) [reply databricks/koalas/data_type_ops/num_ops.py:192]: Wouldn't `column_op(Column.__mul__)(left, right)` be able to handle `NaN`?

I don't see why they cannot reuse `IntegralOps`.

File: databricks/koalas/base.py [hunk 4]
```diff
@@ -229,6 +225,23 @@ def wrapper(self, *args):
         # extract Spark Column. For other arguments, they are used as are.
         cols = [arg for arg in args if isinstance(arg, IndexOpsMixin)]
 
+        # For arithmetic operations involving a bool series, take True as 1, False as 0.
```
Comment (xinrong-databricks) [inline databricks/koalas/base.py:228]: Move to DataTypeOps class

File: databricks/koalas/base.py [hunk 5]
```diff
@@ -229,6 +225,23 @@ def wrapper(self, *args):
...
+        if len(args) == 1:
+            if isinstance(self.spark.data_type, BooleanType):
+                import numbers
+
+                if isinstance(args[0], numbers.Number):
+                    self = self.astype(type(args[0]))
+                elif isinstance(args[0], IndexOpsMixin) and isinstance(
+                    args[0].spark.data_type, NumericType
+                ):
+                    if args[0].dtype != object:
```
Comment (xinrong-databricks) [inline databricks/koalas/base.py:238]: self.spark.apply(lambda scol: scol.cast(args[0].spark.data_type))

File: databricks/koalas/data_type_ops/base.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+            return object.__new__(CategoricalOps)
+        if dtype == np.dtype("object"):
+            if isinstance(spark_type, DecimalType):
+                return object.__new__(DecimalOps)
+            elif isinstance(spark_type, DateType):
+                return object.__new__(DateOps)
+            elif isinstance(spark_type, StringType):
+                return object.__new__(StringOps)
+            else:
+                raise TypeError("Type %s cannot be inferred." % dtype)
```
Comment (xinrong-databricks) [inline databricks/koalas/data_type_ops/base.py:66]: Removed

File: databricks/koalas/data_type_ops/base.py [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+                if not isinstance(right, spark.Column) and pd.isna(right):
+                    return F.lit(False)
+                else:
+                    scol = left | F.lit(right)
+                    return F.when(left.isNull() | scol.isNull(), False).otherwise(scol)
+
+        return column_op(or_func)(left, right)
+
+    def __ror__(self, left, right=None):
+        return self.__or__(left, right)
```
Comment (xinrong-databricks) [inline databricks/koalas/data_type_ops/base.py:198]: Restore and / or operators

File: databricks/koalas/data_type_ops/num_ops.py (hunk 3 above)
Comment (xinrong-databricks) [reply databricks/koalas/data_type_ops/num_ops.py:277]: Removed DecimalOps for now.

**Pull request closed without merging**